### PR TITLE
Curl exporter

### PR DESCRIPTION
**Pull request by instabledesign**

### CODE REVIEW

File: middleware/request_listener_test.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	responseWriter := &httptest.ResponseRecorder{}
+
+	handlerCalled := false
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, req, r)
+		handlerCalled = true
+	})
+
+	called := false
+	mockExporter := func(innerReq *http.Request) {
```
Comment (pimousse1099) [inline middleware/request_listener_test.go]: `listenerMock` ?

File: request_listener/curl.go [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+func CurlLogDumper(request *http.Request) {
+	if request == nil {
+		return
+	}
+	cmd, err := GetCurlCommand(request)
+	if err != nil {
+		log.Println("cannot print curl command", err)
+		return
+	}
+	log.Println(cmd)
```
Comment (pimousse1099) [inline request_listener/curl.go:40]: what if we want to log it in a specific way ? (i.e. with another logger)

File: request_listener/curl.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		if err != nil {
+			return nil, err
+		}
+		req.Body = nopCloser{bytes.NewBuffer(body)}
+		if len(string(body)) > 0 {
+			cmd.append("-d", escape(string(body)))
+		}
+	}
+
+	for h := range req.Header {
```
Comment (pimousse1099) [inline request_listener/curl.go:61]: Maybe handle cookies in another iteratio ? (there are so much options to handle 😨)

File: request_listener/curl.go [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+func GetCurlCommand(req *http.Request) (*Cmd, error) {
+	cmd := &Cmd{
+		"curl",
+		"-X", escape(req.Method),
+		escape(req.URL.String()),
+	}
+
+	if req.Body != nil {
+		body, err := ioutil.ReadAll(req.Body)
```
Comment (pimousse1099) [inline request_listener/curl.go:51]: This is kind of dangerous when the user sends large files don't you think ?

Maybe add a warning in the doc for now ?

**Review by pimousse1099 — CHANGES_REQUESTED** [review]

File: request_listener/curl.go (hunk 2 above)
Comment (instabledesign) [reply request_listener/curl.go:40]: you can:
- implement your own request listener
- use `log.SetOutput(...)`

i want to keep as light as possible for the moment

File: request_listener/curl.go (hunk 4 above)
Comment (instabledesign) [reply request_listener/curl.go:51]: verbose/heavy is more appropriate than `dangerous`.
And all miss used code/library can lead to perf or data lost
This type of data Dumper should be used with some filter.

Yeah i gonna add doc, have you any words to give?

File: request_listener/curl.go [hunk 5]
```diff
@@ -9,6 +9,10 @@ import (
 	"strings"
 )
 
+// CurlLogDumper will log the given http.Request
+// it gonna log the request using cli curl command format
```
Comment (pimousse1099) [inline request_listener/curl.go]: `it will log the request using the `curl` command format`

File: request_listener/curl.go [hunk 6]
```diff
@@ -9,6 +9,10 @@ import (
...
+// Caution request header and Body can be heavy!!
```
Comment (pimousse1099) [inline request_listener/curl.go]: `/!\ Caution request header and Body can be heavy !!` ?

**Review by pimousse1099 — APPROVED** [review]

File: request_listener/curl.go (hunk 2 above)
Comment (pimousse1099) [reply request_listener/curl.go:40]: OK I see 👍

**Review by pimousse1099 — APPROVED** [review]

File: request_listener/curl.go [hunk 7]
```diff
@@ -9,10 +9,23 @@ import (
 	"strings"
 )
 
-// CurlLogDumper will log the given http.Request
-// it gonna log the request using cli curl command format
-// Caution request header and Body can be heavy!!
-// Pay attention to the dump quantity/weight
+// CurlLogDumper will log the request using the curl command format
+// /!\ Caution request header and Body can be heavy !!
+// Eg tripperware:
```
Comment (pimousse1099) [inline request_listener/curl.go:16]: Maybe add something like this:

`You can use this tripperware to only dump the request if `dump` HTTP Header is present` ?

**Review by pimousse1099 — APPROVED** [review]

**Pull request merged**